heat/core/manipulations.py

```diff
@@ -1331,6 +1426,182 @@ def squeeze(x, axis=None):
+             [15, 16, 17, 18, 19]],
+
+            [[20, 21, 22, 23, 24],
+             [25, 26, 27, 28, 29],
+             [30, 31, 32, 33, 34],
+             [35, 36, 37, 38, 39]]])
+    # distributed DNDarrays, 3 processes, stack along last dimension
+    >>> a = ht.arange(20, split=0).reshape(4, 5)
+    >>> b = ht.arange(20, 40, split=0).reshape(4, 5)
+    >>> ht.stack((a,b), axis=-1)
+    [0/2] tensor([[[ 0, 20],
+    [0/2]          [ 1, 21],
+    [0/2]          [ 2, 22],
+    [0/2]          [ 3, 23],
+    [0/2]          [ 4, 24]],
+    [0/2]
+    [0/2]         [[ 5, 25],
+    [0/2]          [ 6, 26],
+    [0/2]          [ 7, 27],
+    [0/2]          [ 8, 28],
+    [0/2]          [ 9, 29]]])
+    [1/2] tensor([[[10, 30],
+    [1/2]          [11, 31],
+    [1/2]          [12, 32],
+    [1/2]          [13, 33],
+    [1/2]          [14, 34]]])
+    [2/2] tensor([[[15, 35],
+    [2/2]          [16, 36],
+    [2/2]          [17, 37],
+    [2/2]          [18, 38],
+    [2/2]          [19, 39]]])
+    """
+
+    # sanitation
+    if not isinstance(arrays, (tuple, list)):
+        raise TypeError("arrays must be a list or a tuple")
+
+    if len(arrays) < 2:
+        raise ValueError("stack expects a sequence of at least 2 DNDarrays")
+
+    for i, array in enumerate(arrays):
+        if not isinstance(array, dndarray.DNDarray):
+            raise TypeError(
+                "all arrays in sequence must be DNDarrays, array {} was {}".format(i, type(array))
+            )
+
+    arrays_metadata = list(
+        [array.gshape, array.split, array.device, array.is_balanced()] for array in arrays
+    )
+    num_arrays = len(arrays)
+    # metadata must be identical for all arrays
+    if arrays_metadata.count(arrays_metadata[0]) != num_arrays:
+        shapes = list(array.gshape for array in arrays)
+        if shapes.count(shapes[0]) != num_arrays:
+            raise ValueError(
+                "All DNDarrays in sequence must have the same shape, got shapes {}".format(shapes)
+            )
+        splits = list(array.split for array in arrays)
+        if splits.count(splits[0]) != num_arrays:
+            raise ValueError(
+                "All DNDarrays in sequence must have the same split axis, got splits {}"
+                "Check out the heat.resplit() documentation.".format(splits)
+            )
+        devices = list(array.device for array in arrays)
+        if devices.count(devices[0]) != num_arrays:
+            raise RuntimeError(
+                "DNDarrays in sequence must reside on the same device, got devices {}".format(
+                    devices
+                )
+            )
+        balance = list(array.is_balanced() for array in arrays)
+        if balance.count(balance[0]) != num_arrays:
+            raise RuntimeError(
+                "DNDarrays distribution must be balanced across ranks, is_balanced() returns {}"
+                "You can balance a DNDarray with the balance_() method.".format(balance)
+            )
```

Comment:
why is there a loop at 1520? couldnt this be removed entirely? then it would check the shapes, splits, devices, and balance for all of the inputs?

Reply:
> why is there a loop at 1520? couldnt this be removed entirely? then it would check the shapes, splits, devices, and balance for all of the inputs?

It could. But checking the metadata one by one is only relevant when we need to return an error message, and  in the vast majority of cases, users will have their arrays set up just right. Why do we have to go through 4 checks before we even start, when we can be done in one quick coarse check? 

This will be superseeded by the sanitation module anyway, by the way.

Reply:
ah okay, i misunderstood the count. then it looks good to me